### PR TITLE
Fix #4445: Integrate transaction confirmation UI with tx controller

### LINKED ISSUE #4445: Integrate Brave Wallet Transaction Confirmation UI with API
Integrate the wallet transaction confirmation screen with `approveTransaction` API

### PR DESCRIPTION
**Pull request by kylehickinson**
## Summary of Changes

This pull request fixes #4445

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift [hunk 1]
```diff
@@ -21,21 +24,92 @@ struct TransactionConfirmationView: View {
...
-    if activeTransactionIndex == transactions.endIndex {
-      activeTransactionIndex = 0
+    if let index = transactions.firstIndex(where: { $0.id == activeTransactionId }) {
+      var nextIndex = transactions.index(after: index)
+      if nextIndex == transactions.endIndex {
+        nextIndex = 0
+      }
+      activeTransactionId = transactions[nextIndex].id
+    } else {
+      activeTransactionId = transactions.first!.id
```
Comment (iccub) [inline BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift:41]: can this be empty at any point?

File: BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift [hunk 2]
```diff
@@ -21,21 +24,92 @@ struct TransactionConfirmationView: View {
...
   private func rejectAll() {
-    presentationMode.dismiss()
+    for transaction in transactions {
+      confirmationStore.reject(transaction: transaction)
+    }
   }
   
   private var activeTransaction: BraveWallet.TransactionInfo {
-    transactions[activeTransactionIndex]
+    transactions.first(where: { $0.id == activeTransactionId }) ?? transactions.first!
```
Comment (iccub) [inline BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift:52]: same question

File: BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift [hunk 3]
```diff
@@ -21,21 +24,92 @@ struct TransactionConfirmationView: View {
...
+    default:
+      if activeTransaction.isSwap {
+        return Strings.Wallet.swap
+      }
+      return Strings.Wallet.send
+    }
+  }
+  
+  private var transactionDetails: String {
+    let data = activeTransaction.txData.baseData.data
```
Comment (iccub) [inline BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift:76]: Would make sense to make a helper function for that I think, is data to `0x` hex string conversion used anywhere else in the app?

File: BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift [hunk 4]
```diff
@@ -21,21 +24,92 @@ struct TransactionConfirmationView: View {
...
+      .map { byte in
+        String(format: "%02X", byte.uint8Value)
```
Comment (iccub) [inline BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift:78]: Would make sense to make a helper function 'DataExtensions' for that I think, is data to `0x` string conversion used anywhere else in the app?

File: BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift (hunk 1 above)
Comment (kylehickinson) [reply BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift:41]: No this screen can never be presented unless transactions is non-empty. It is also guarded against this here: https://github.com/brave/brave-ios/pull/4567/files#diff-b5e46895fcabbaf5ddbc0680da3204666fcf76f65883af6f7a4e3e265a5ead3cR97

File: BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift (hunk 3 above)
Comment (kylehickinson) [reply BraveWallet/Crypto/Transaction Confirmations/TransactionConfirmationView.swift:76]: Don't believe we have anything else in wallet that uses an array of NSNumber/uint8's. I think this is specifically a mojo limitation

File: BraveWallet/Crypto/CryptoView.swift [hunk 5]
```diff
@@ -11,7 +11,7 @@ import Introspect
 import BraveUI
 
 public struct CryptoView: View {
-  var walletStore: WalletStore
+  @ObservedObject var walletStore: WalletStore
   @ObservedObject var keyringStore: KeyringStore
```
Comment (nuo-xu) [inline BraveWallet/Crypto/CryptoView.swift:16]: if walletStore is also @ObservedObject will a change in `keyringStore` also a change in `walletStore`?

Comment (kylehickinson) [reply BraveWallet/Crypto/CryptoView.swift:16]: Nope, `ObservableObject` does not hook up any child `ObservableObject ` by default, you have to do so yourself if you want that behaviour

File: BraveWallet/Crypto/Stores/TransactionConfirmationStore.swift [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    let gasFee: String
+    if transaction.isEIP1559Transaction {
+      gasFee = transaction.txData.maxFeePerGas
+    } else {
+      gasFee = transaction.txData.baseData.gasPrice
+    }
+    self.state.gasValue = gasFormatter.decimalString(
+      for: gasFee.removingHexPrefix,
+         radix: .hex,
+         decimals: Int(selectedChain.decimals)
```
Comment (nuo-xu) [inline BraveWallet/Crypto/Stores/TransactionConfirmationStore.swift:65]: indentation is weird here.

Comment (nuo-xu) [reply BraveWallet/Crypto/Stores/TransactionConfirmationStore.swift:65]: I believe txData.maxFeePerGas is in GWei. do we need to give decimals: Int(selectedChain.decimals) which i think will make it in Wei?

Comment (kylehickinson) [reply BraveWallet/Crypto/Stores/TransactionConfirmationStore.swift:65]: Yeah thats Xcode :( Can't do anything about it unfortunately

Comment (kylehickinson) [reply BraveWallet/Crypto/Stores/TransactionConfirmationStore.swift:65]: I tried it with 9 decimals and got the wrong values so I guess it is stored in Wei, but displayed in Gwei

File: BraveWallet/Crypto/Stores/TransactionConfirmationStore.swift [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  func fetchDetails(for transaction: BraveWallet.TransactionInfo) {
+    state = .init() // Reset state
+    isLoading = true
+    activeTransaction = transaction
+    
+    rpcController.chainId { [weak self] chainId in
+      guard let self = self else { return }
+      self.rpcController.allNetworks { chains in
+        let selectedChain = chains.first(where: { $0.chainId == chainId }) ?? .init()
+        self.selectedChain = selectedChain
```
Comment (nuo-xu) [inline BraveWallet/Crypto/Stores/TransactionConfirmationStore.swift]: we can now use rpcController.network

**Review by nuo-xu — APPROVED** [review]

**Pull request merged**